File: src/main/java/org/example/expert/controller/manager/ManagerController.java
package org.example.expert.controller.manager;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.example.expert.config.security.loginuser.LoginUser;
import org.example.expert.controller.manager.dto.request.ManagerSaveRequest;
import org.example.expert.controller.manager.dto.response.ManagerResponse;
import org.example.expert.controller.manager.dto.response.ManagerSaveResponse;
import org.example.expert.service.ManagerService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequiredArgsConstructor
public class ManagerController {

    private final ManagerService managerService;

    @PostMapping("/todos/{todoId}/managers")
    public ResponseEntity<ManagerSaveResponse> saveManager(
            @AuthenticationPrincipal LoginUser loginUser,
            @PathVariable(value = "todoId") Long todoId,
            @Valid @RequestBody ManagerSaveRequest managerSaveRequest
    ) {
        return ResponseEntity.ok(managerService.saveManager(loginUser.getUser(), todoId, managerSaveRequest));
    }

    @GetMapping("/todos/{todoId}/managers")
    public ResponseEntity<List<ManagerResponse>> getMembers(@PathVariable(value = "todoId") Long todoId) {
        return ResponseEntity.ok(managerService.getManagers(todoId));
    }

    @DeleteMapping("/todos/{todoId}/managers/{managerId}")
    public void deleteManager(
            @AuthenticationPrincipal LoginUser loginUser,
            @PathVariable(value = "todoId") Long todoId,
            @PathVariable(value = "managerId") Long managerId
    ) {
        managerService.deleteManager(loginUser.getUser(), todoId, managerId);
    }
}

File: src/main/java/org/example/expert/controller/manager/ManagerController.kt
package org.example.expert.controller.manager

import jakarta.validation.Valid
import org.example.expert.config.security.loginuser.LoginUser
import org.example.expert.controller.manager.dto.request.ManagerCreateReqDto
import org.example.expert.controller.manager.dto.response.ManagerCreateRespDto
import org.example.expert.controller.manager.dto.response.ManagerInfoRespDto
import org.example.expert.service.ManagerService
import org.example.expert.util.api.ApiResult
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/todos")
class ManagerController(
    private val managerService: ManagerService
) {

    @PostMapping("/{todoId}/managers")
    fun saveManager(@Valid @RequestBody managerSaveReqDto: ManagerCreateReqDto,
                    @PathVariable(value = "todoId") todoId: Long,
                    @AuthenticationPrincipal loginUser: LoginUser)
    : ResponseEntity<ApiResult<ManagerCreateRespDto>> =
        ResponseEntity.ok(ApiResult.success(managerService.saveManager(loginUser.user, todoId, managerSaveReqDto)))


    @GetMapping("/{todoId}/managers")
    fun getManagerList(@PathVariable(value = "todoId")todoId: Long)
    : ResponseEntity<ApiResult<List<ManagerInfoRespDto>>> =
        ResponseEntity.ok(ApiResult.success(managerService.getManagerList(todoId)))


    @DeleteMapping("/{todoId}/managers/{managerId}")
    fun deleteManager(@PathVariable(value = "todoId")todoId: Long,
                      @PathVariable(value = "managerId")managerId: Long,
                      @AuthenticationPrincipal loginUser: LoginUser)
    : ResponseEntity<ApiResult<String>> {
        managerService.deleteManager(loginUser.user, todoId, managerId)
        return ResponseEntity.ok(ApiResult.success("해당 담당자가 성공적으로 삭제되었습니다"))
    }

}